bigchaindb/common/transaction.py

```diff
@@ -575,10 +575,17 @@ def __init__(self, operation, asset, fulfillments=None, conditions=None,
             raise ValueError('`operation` must be one of {}'
                              .format(allowed_ops))
 
-        # Only assets for 'CREATE' operations can be un-defined.
-        if (asset and not isinstance(asset, Asset) or
-                not asset and operation != Transaction.CREATE):
-            raise TypeError('`asset` must be an Asset instance')
+        # Assets for 'CREATE' and 'GENESIS' operations must be None or of Asset
+        # type and Assets for 'TRANSFER' operations must be of AssetLink type.
+        if (operation in [Transaction.CREATE, Transaction.GENESIS] and
+                asset is not None and
+                not isinstance(asset, Asset)):
+            raise TypeError(("`asset` must be an Asset instance for "
+                             "'{}' Transactions".format(operation)))
+        elif (operation == Transaction.TRANSFER and
```

Comment:
This is a big breaking change. `AssetLink` was something that only add meaning in the server code.
Clients would always pass an asset regardless and we would serialized it according to the operation.

Its probably worth it to open an issue in the bigchaindb-driver